### DISCUSSION TITLE
Change lines - quick question

### DISCUSSION THREAD

**Discussion by seyeint** [highfestiva/finplot#447] [Q&A] [answered]
AttributeError: 'FinLine' object has no attribute 'setPoints'

I have a list of lines that I am using like this whenever I want to draw something because I know I will need to change them in the future, meaning I will have to access them:
            entry_lines.append(fplt.add_line((...))
So the problem is I want to update them and I found that you have this function "points()" that returns the tuples of both points of a line (it does the x relative to the number of candles i guess and not absolute timeframe but no problem) and then you also have "setPoints()" to change them. I have no problem accessing the lines from my list but when I try to do setPoints, it tells me FinLine has no such attribute:

        for j in entry_lines:
            old_p1, old_p2 = j.points
            j.setPoints(old_p1, (old_p2[0]+1, old_p2[1]))

**Comment by seyeint** [comment]
j.points = (old_p1, (old_p2[0]+1, old_p2[1]))
this worked

**Comment by seyeint** [comment]
is there a way to use the absolute timestamp and not a real number of candles as the x tuple? it's messing with my other timeframes when i change, as there are more or less candles, hence the lines change place

**Reply by highfestiva** [reply]
I think there's a helper function within finplot for that, but you could also roll your own to transform back and fro.

**Reply by seyeint** [reply]
thanks for the answer, btw by "helper function" what do you mean exactly? where can i access that?

**Reply by highfestiva** [reply]
There are two. You may use them, and they probably won't change in future versions, but just know that if they do you'll have to update your code. The two are `_x2local_t(datasrc, x)` and `_x2utc(datasrc, x)`. Call something like this: `fplt._x2utc(ax.vb.datasrc, my_x_coord)`. GL!

**Reply by seyeint** [reply]
Ok I tried and seen what that did. So that function helps me retrieve the date of my line.points result.
So for example I isolated one of two points that make a line j, lets call it p1 and p2. 

**old_p1, old_p2 = j.points
print(old_p1,fplt._x2utc(ax.vb.datasrc, old_p1[0]))**
j.points = (old_p1, (old_p2[0]+1, old_p2[1]))

which gives me:
(**_19_**, 10778.0) ('**_2020-09-27 04:45_**', True)

Ok but the problem is that what I want is to update/set new coordinates to that same line object without creation a new one, which i am doing on my last line where i do j.points=(p1, p2+1candle) ---> but I need to update it with timestamp date values because if I continue with **"19"** or 19+1 when moving forward in time, the line only stays in the correct place in that same dataframe. The moment i switch my view in my program to a 1h from lets say 15min, the line teleports forward instead of staying in the correct date (from the 15minutes), because it is counting number of candles and obviously 4h has less candles than 15min to represent a static time gap.

 i need something like j.points = (old_p1, (1601181000000000000, old_p2[1])), assuming 1601181000000000000 is where i want to stretch the line towards.

and by the way, even without updating the line to different x - dates, the moment i change the timeframe it already changes place although i thought it wouldn't be a problem since in the actual add line function fplt.add_line((order_date, int(price)), (t_shown_max * (10 ** 9), int(price)), color='#8db290', interactive=False), the order date and t_show_max are actual timestamps like the one i exemplified in the previous paragraph, which would indicate to me that they would be drawn correctly on multiple different timeframes assuming i wouldn't start messing with them, by updating them with the method i know uses number of candles as x - time.

**Reply by seyeint** [reply, marked as answer]
Final answer, **its closed now.**
The function add_line ALWAYS transforms the raw date into ordinal/index time relative to the number of candles in the current chart.
Meaning one can CREATE lines with normal timestamp dates, but if one wants to update and not just eliminate and re-draw, one has to now use the following method to get their x coordinates for the line.points() update.


py = price where that line was innitially marked.
p1x, p2x = fplt._pdtime2index(ax, pd.Series([p1x, p2x])), where p1x can be for example the old p1x and p2x might be the new extended p2x.
old_line.points = ((p1x, py), (p2x, py))
